File: kotlin-android/app/src/main/java/computer/iroh/dot/net/IrohPeer.kt
package computer.iroh.dot.net

import android.os.Build
import android.util.Log
import computer.iroh.BiStream
import computer.iroh.Connection
import computer.iroh.Endpoint
import computer.iroh.EndpointAddr
import computer.iroh.EndpointId
import computer.iroh.EndpointOptions
import computer.iroh.ServicesClient
import computer.iroh.ServicesOptions
import computer.iroh.presetN0
import computer.iroh.dot.game.DotGame
import computer.iroh.dot.game.MotionSource
import computer.iroh.dot.identity.IdentityStore
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

private const val TAG = "IrohPeer"

class IrohPeer(
    private val scope: CoroutineScope,
    private val motion: MotionSource,
    private val identity: IdentityStore,
) {
    sealed interface State {
        data object Idle : State
        data object Binding : State
        data object Ready : State
        data object Connecting : State
        data class Connected(val peerShortId: String) : State
        data class Error(val message: String) : State
    }

    val game = DotGame()

    private val _state = MutableStateFlow<State>(State.Idle)
    val state: StateFlow<State> = _state.asStateFlow()

    private val _endpointId = MutableStateFlow(identity.endpointId)
    val endpointId: StateFlow<String> = _endpointId.asStateFlow()

    /**
     * One-line live view of the connection's open network paths
     * (direct vs relay, address, RTT), rendered under the status line.
     */
    private val _pathInfo = MutableStateFlow("")
    val pathInfo: StateFlow<String> = _pathInfo.asStateFlow()

    private var endpoint: Endpoint? = null
    private var acceptJob: Job? = null
    private var pathMonitorJob: Job? = null
    private var session: PeerSession? = null
    private var services: ServicesClient? = null

    fun start() {
        if (endpoint != null) return
        _state.value = State.Binding
        scope.launch {
            try {
                val ep = Endpoint.bind(
                    EndpointOptions(
                        preset = presetN0(),
                        secretKey = identity.secretKey.toBytes(),
                        alpns = listOf(WireFormat.ALPN),
                    ),
                )
                endpoint = ep
                _state.value = State.Ready
                acceptJob = scope.launch { runAcceptLoop(ep) }
                startServicesClient()
            } catch (t: Throwable) {
                _state.value = State.Error("bind failed: ${t.message ?: t}")
            }
        }
    }

    fun connect(toEndpointIdHex: String) {
        val ep = endpoint ?: return
        val trimmed = toEndpointIdHex.trim()
        scope.launch {
            val parsed = try {
                EndpointId.fromString(trimmed)
            } catch (t: Throwable) {
                _state.value = State.Error("invalid endpoint id")
                return@launch
            }
            _state.value = State.Connecting
            try {
                val addr = EndpointAddr(parsed, null, emptyList())
                val conn = ep.connect(addr, WireFormat.ALPN)
                val bi = conn.openBi()
                adoptSession(bi, conn, parsed.toString())
            } catch (t: Throwable) {
                _state.value = State.Error("connect failed: ${t.message ?: t}")
            }
        }
    }

    private suspend fun startServicesClient() {
        val ep = endpoint ?: return
        try {
            services = ServicesClient.create(
                ep,
                ServicesOptions(apiSecret = API_SECRET, name = deviceName()),
            )
        } catch (t: Throwable) {
            Log.w(TAG, "iroh services client not started: ${t.message ?: t}")
        }
    }

    private fun deviceName(): String {
        val short = _endpointId.value.take(8)
        val model = Build.MODEL.lowercase().replace(" ", "-")
        return "android-$model-$short"
    }

    private suspend fun runAcceptLoop(ep: Endpoint) {
        while (currentCoroutineContext().isActive) {
            val incoming = try {
                ep.acceptNext() ?: return
            } catch (t: Throwable) {
                Log.w(TAG, "acceptNext threw", t)
                continue
            }
            try {
                val accepting = incoming.accept()
                val alpn = accepting.alpn()
                if (!alpn.contentEquals(WireFormat.ALPN)) continue
                val conn = accepting.connect()
                val bi = conn.acceptBi()
                val remoteIdHex = conn.remoteId().toString()
                adoptSession(bi, conn, remoteIdHex)
            } catch (t: Throwable) {
                Log.w(TAG, "accept flow threw", t)
            }
        }
    }

    private fun adoptSession(bi: BiStream, conn: Connection, remoteIdHex: String) {
        session?.stop()
        game.resetForNewSession()
        startPathMonitor(conn)
        val g = game
        val m = motion
        lateinit var s: PeerSession
        s = PeerSession(
            bi = bi,
            produceFrame = {
                g.setMyPos(m.x, m.y)
                WireFormat.encodePosition(g.myPos.x, g.myPos.y)
            },
            onPositionReceived = { p -> g.receivedTheirPos(p.x, p.y) },
            // Capture the session identity in the closure so a late-firing
            // onClosed from a previous session can't tear down the current one.
            onClosed = { if (session === s) handleSessionClosed() },
        )
        session = s
        s.start(scope)
        val short = remoteIdHex.take(10)
        _state.value = State.Connected(short)
        Log.d(TAG, "session adopted with $remoteIdHex")
    }

    private fun handleSessionClosed() {
        session = null
        pathMonitorJob?.cancel()
        pathMonitorJob = null
        _pathInfo.value = ""
        game.sessionEnded()
        if (_state.value is State.Connected) _state.value = State.Ready
    }

    /**
     * Polls the connection's path snapshots once a second and renders them
     * as one line, so direct-vs-relay and RTT are visible in the UI. The
     * selected path (the one carrying application data) is marked with "*".
     */
    private fun startPathMonitor(conn: Connection) {
        pathMonitorJob?.cancel()
        pathMonitorJob = scope.launch {
            while (currentCoroutineContext().isActive) {
                _pathInfo.value = conn.paths().joinToString(" | ") { p ->
                    val kind = if (p.isRelay) "relay" else if (p.isIp) "direct" else "?"
                    val mark = if (p.isSelected) "* " else ""
                    "$mark$kind ${p.remoteAddr} ${p.rttMs}ms"
                }
                delay(1_000)
            }
        }
    }

    companion object {
        // Paste an API key from https://services.iroh.computer to see this
        // device's metrics in your dashboard. With the placeholder left in
        // place the services client fails to start; the demo works either way.
        const val API_SECRET = "<your-iroh-services-api-key>"
    }
}

File: swift/HelloIroh/IrohPeer.swift
import Foundation
import Observation
import IrohLib
#if canImport(UIKit)
import UIKit
#elseif canImport(AppKit)
import AppKit
#endif

@MainActor
@Observable
final class IrohPeer {
    enum ConnectionState: Equatable {
        case idle
        case binding
        case ready
        case connecting
        case connected(peerShortId: String)
        case error(String)
    }

    let motion: MotionSource
    let game = DotGame()

    var endpointId: String = ""
    var state: ConnectionState = .idle

    /// Paste an API key from https://services.iroh.computer to see this
    /// device's metrics in your dashboard. With the placeholder left in
    /// place the services client fails to start; the demo works either way.
    static let apiSecret = "<your-iroh-services-api-key>"

    private var endpoint: Endpoint?
    private var acceptTask: Task<Void, Never>?
    private var currentSession: PeerSession?
    private var identity: IdentityStore?
    private var services: ServicesClient?

    init(motion: MotionSource) {
        self.motion = motion
    }

    func start() async {
        guard endpoint == nil else { return }
        state = .binding
        let identity = IdentityStore.loadOrCreate()
        self.identity = identity
        endpointId = identity.endpointId
        do {
            let ep = try await Endpoint.bind(options: EndpointOptions(
                preset: presetN0(),
                secretKey: identity.secretKey.toBytes(),
                alpns: [WireFormat.alpn]
            ))
            endpoint = ep
            state = .ready
            acceptTask = Task { [weak self] in
                await self?.runAcceptLoop(ep)
            }
            await startServicesClient()
        } catch {
            state = .error("bind failed: \(error)")
        }
    }

    private func startServicesClient() async {
        guard let ep = endpoint else { return }
        do {
            services = try await ServicesClient.create(
                endpoint: ep,
                options: ServicesOptions(apiSecret: Self.apiSecret, name: deviceName())
            )
        } catch {
            print("iroh services client not started: \(error)")
        }
    }

    private func deviceName() -> String {
        let short = String(endpointId.prefix(8))
        #if os(iOS)
        return "ios-\(short)"
        #elseif os(macOS)
        return "macos-\(short)"
        #else
        return "dot-\(short)"
        #endif
    }

    func connect(toEndpointIdHex hex: String) async {
        guard let ep = endpoint else { return }
        let trimmed = hex.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let parsed = try? EndpointId.fromString(s: trimmed) else {
            state = .error("invalid endpoint id")
            return
        }
        state = .connecting
        let addr = EndpointAddr(id: parsed, relayUrl: nil, addresses: [])
        do {
            let conn = try await ep.connect(addr: addr, alpn: WireFormat.alpn)
            let bi = try await conn.openBi()
            adoptSession(bi: bi, remoteIdHex: parsed.description)
        } catch {
            state = .error("connect failed: \(error)")
        }
    }

    func copyEndpointIdToClipboard() {
        #if canImport(UIKit)
        UIPasteboard.general.string = endpointId
        #elseif canImport(AppKit)
        NSPasteboard.general.clearContents()
        NSPasteboard.general.setString(endpointId, forType: .string)
        #endif
    }

    private func runAcceptLoop(_ ep: Endpoint) async {
        while !Task.isCancelled {
            guard let incoming = await ep.acceptNext() else { return }
            do {
                let accepting = try await incoming.accept()
                let alpn = try await accepting.alpn()
                guard alpn == WireFormat.alpn else { continue }
                let conn = try await accepting.connect()
                let bi = try await conn.acceptBi()
                let remoteId = conn.remoteId().description
                adoptSession(bi: bi, remoteIdHex: remoteId)
            } catch {
                continue
            }
        }
    }

    private func adoptSession(bi: BiStream, remoteIdHex: String) {
        currentSession?.stop()
        game.resetForNewSession()
        let game = self.game
        let motion = self.motion
        let session = PeerSession(
            bi: bi,
            produceFrame: {
                await MainActor.run {
                    game.setMyPos(x: motion.x, y: motion.y)
                    return WireFormat.encodePosition(x: game.myPos.x, y: game.myPos.y)
                }
            },
            onPositionReceived: { pos in
                Task { @MainActor in game.receivedTheirPos(x: pos.x, y: pos.y) }
            },
            onClosed: { [weak self] in
                Task { @MainActor in self?.handleSessionClosed() }
            }
        )
        currentSession = session
        session.start()
        let short = String(remoteIdHex.prefix(10))
        state = .connected(peerShortId: short)
    }

    private func handleSessionClosed() {
        guard currentSession != nil else { return }
        currentSession = nil
        game.sessionEnded()
        if case .connected = state { state = .ready }
    }
}
